### PR TITLE
Fixed compilation with cmake on Windows

### LINKED ISSUE #16161: [Feature request] Allow RPCS3 to be compiled in DEBUG mode
### Quick summary

In order to instrument the code with debugging tools such as valgrind (e.g. under linux), Deleaker or even VS 2022 debugger it should be very useful to review the project allowing the compilation in DEBUG mode. Currently, most of the linking errors are on qt libs (and that should be fixable). We know that some libs could be affected by bugs in DEBUG mode but the main target of this FR is to allow to discover clear issues in RPCS3 source code. Currently this is not possible at all. Debugging the code in RELEASE mode is pretty much useless (I see an infinite number of access violation, they can be real or false issues but it is impossible to understand that not having a reference to source code for a further analysis)

### Details

_No response_

### PR DESCRIPTION
**Pull request by digant73**
- Fixed compilation errors with `cmake` on Windows OS
- Updated `BUILDING.md` file (sections related to Windows)

Credits to @oltolm for the support here and on #16161

fixes #16161

**PR STATUS:** Ready for review and merge

### CODE REVIEW

File: 3rdparty/llvm/CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,8 @@ if(WITH_LLVM)
 		option(LLVM_CCACHE_BUILD OFF)
 		set(LLVM_ENABLE_WARNINGS OFF CACHE BOOL "Enable compiler warnings.")
 
-		if(WIN32 AND COMPILER_X86)
+		# For non arm64 windows
+		if(WIN32 AND (COMPILER_X86 OR (MSVC AND NOT COMPILER_ARM)))
```
Comment (kd-11) [inline 3rdparty/llvm/CMakeLists.txt:21]: MSVC-amd64 should set COMPILER_X86. Is this not happening right now?

File: CMakePresets.json [hunk 2]
```diff
@@ -48,13 +48,16 @@
         "strategy": "external"
       },
       "cacheVariables": {
+        "CMAKE_BUILD_TYPE": "Debug",
```
Comment (oltolm) [inline CMakePresets.json:51]: Visual Studio can switch the build type in the IDE or with `cmake --build . --config Debug`. It's not correct to add it here.

File: BUILDING.md [hunk 3]
```diff
@@ -7,25 +7,41 @@ Other instructions may be found [here](https://wiki.rpcs3.net/index.php?title=Bu
...
+    with standalone **CMake** tool.
 
-**NOTE: If you have issues with the Qt plugin, you may want to uninstall the Qt Plugin and install the [Legacy Qt Plugin](https://marketplace.visualstudio.com/items?itemName=TheQtCompany.LEGACYQtVisualStudioTools2019) instead.**
+- [Python 3.6+](https://www.python.org/downloads/) (add to PATH)
+- [Qt 6.7.3](https://www.qt.io/download-qt-installer)
+- [Vulkan SDK 1.3.268.0](https://vulkan.lunarg.com/sdk/home) (see "Install the SDK" [here](https://vulkan.lunarg.com/doc/sdk/latest/windows/getting_started.html)) for now future SDKs don't work. You need precisely 1.3.268.0.
 
-### Linux
+The `sln` solution available only on **Visual Studio** is the preferred building solution. It allows to build the **RPCS3** application in `Release` and `Debug` mode
+while the `CMake` solution is currently limited to `Debug` mode only.
```
Comment (oltolm) [inline BUILDING.md:26]: That's not correct. CMake and Visual Studio support all configurations.

File: rpcs3/CMakeLists.txt [hunk 4]
```diff
@@ -153,7 +153,9 @@ elseif(UNIX)
             ${CMAKE_SOURCE_DIR}/bin/git $<TARGET_FILE_DIR:rpcs3>/git)
 elseif(WIN32)
     add_custom_command(TARGET rpcs3 POST_BUILD
+        COMMAND "${CMAKE_COMMAND}" -E remove "${CMAKE_SOURCE_DIR}/bin/rpcs3.exe"
```
Comment (oltolm) [inline rpcs3/CMakeLists.txt:156]: Why is this necessary?

File: rpcs3/CMakeLists.txt [hunk 5]
```diff
@@ -153,7 +153,9 @@ elseif(UNIX)
...
         COMMAND "${CMAKE_COMMAND}" -E copy_directory "${CMAKE_SOURCE_DIR}/bin" "$<TARGET_FILE_DIR:rpcs3>"
+        COMMAND "${CMAKE_COMMAND}" -E copy "${PROJECT_BINARY_DIR}/3rdparty/OpenAL/openal-soft/Debug/OpenAL32.dll" "$<TARGET_FILE_DIR:rpcs3>"
```
Comment (oltolm) [inline rpcs3/CMakeLists.txt:158]: This will only work with the Debug config and MSVC, but we support other configs and GCC, Clang on Windows.

**Review by oltolm — CHANGES_REQUESTED** [review]

File: rpcs3/CMakeLists.txt (hunk 4 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt:156]: because the post build process replaced the compiled `rpcs3.exe` with the one present in `rpcs3/bin`.
I would prefer something like `EXCLUDE rpcs3.exe` in the next command (with the command `-E copy_directory`) but I didn't find any option on that command.

File: rpcs3/CMakeLists.txt (hunk 5 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt:158]: I can apply those changes only for MSVC if needed. The problem is that on Windows (at least on my 7800X3D) the `Release` mode had no effect at all. The application is always compiled in `Debug` mode from both VS and also standalone `cmake` (from command line).

File: CMakePresets.json (hunk 2 above)
Comment (digant73) [reply CMakePresets.json:51]: I know but on my Windows 11 with 7800X3D the `Release` mode has no effect at all. The application is always compiled in `Debug` mode. Same result compiling with both VS and the standalone cmake (external cmake tool)

File: 3rdparty/llvm/CMakeLists.txt (hunk 1 above)
Comment (oltolm) [reply 3rdparty/llvm/CMakeLists.txt:21]: It is better to change the lines 73-78 to
```
		if(LLVM_USE_INTEL_JITEVENTS)
			list (APPEND LLVM_ADDITIONAL_LIBS IntelJITEvents)
		endif()
		if(LLVM_USE_PERF)
			list (APPEND LLVM_ADDITIONAL_LIBS PerfJITEvents)
		endif()
```
I tested it and it worked.

File: CMakePresets.json (hunk 2 above)
Comment (oltolm) [reply CMakePresets.json:51]: `CMAKE_BUILD_TYPE` does not work for multi-config generators.

From the official CMake documentation.
> For multi-configuration generators like [Visual Studio](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#visual-studio-generators), [Xcode](https://cmake.org/cmake/help/latest/generator/Xcode.html#generator:Xcode), and [Ninja Multi-Config](https://cmake.org/cmake/help/latest/generator/Ninja%20Multi-Config.html#generator:Ninja%20Multi-Config), the configuration is chosen by the user at build time and [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html#variable:CMAKE_BUILD_TYPE) is ignored.

I don't know what you mean when say it had no effect. I just tested it and it worked as expected.

Comment (digant73) [reply CMakePresets.json:51]: I mean that from the IDE I cannot select the compile config `Release` or `Debug` (even removing `"CMAKE_BUILD_TYPE": "Debug"` from `CMakePresets.json`). If I select `rpcs3.exe` from the Solution Explorer panel there is no possibility to select  `Release` or `Debug`. See attached picture.

vs_cmake

If I provide `"CMAKE_BUILD_TYPE": "Debug"` or `"CMAKE_BUILD_TYPE": Release"` (as it is done for `gcc` and `clang`) the application is always compiled in `Debug`. From command line, I will verify  the suggested `--config Debug` option

File: rpcs3/CMakeLists.txt (hunk 5 above)
Comment (Megamouse) [reply rpcs3/CMakeLists.txt:158]: I don't think check_cxx_compiler_flag is called for MSVC.
I don't know anything about cmake, but this stood out to me.

Comment (Megamouse) [reply rpcs3/CMakeLists.txt:158]: maybe it doesn't have the CMPXCHG16B instruction

Comment (oltolm) [reply rpcs3/CMakeLists.txt:158]: That's from the CMake log
```
        ClCompile:
          C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\CL.exe /c /I"C:\\src\\vcpkg\\installed\\x64-windows\\include" /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /D _MBCS /D WIN32 /D _WINDOWS /D COMPILER_X86 /D "CMAKE_INTDIR=\\"Debug\\"" /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++20 /Fo"cmTC_a7f2c.dir\\Debug\\\\" /Fd"cmTC_a7f2c.dir\\Debug\\vc143.pdb" /external:W1 /Gd /TP /errorReport:queue  -msse -msse2 -mcx16 "E:\\build-rpcs3-msvc\\CMakeFiles\\CMakeScratch\\TryCompile-szusic\\src.cxx"
          Microsoft (R) C/C++ Optimizing Compiler Version 19.41.34120 for x64
          Copyright (C) Microsoft Corporation.  All rights reserved.
          cl /c /I"C:\\src\\vcpkg\\installed\\x64-windows\\include" /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /D _MBCS /D WIN32 /D _WINDOWS /D COMPILER_X86 /D "CMAKE_INTDIR=\\"Debug\\"" /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++20 /Fo"cmTC_a7f2c.dir\\Debug\\\\" /Fd"cmTC_a7f2c.dir\\Debug\\vc143.pdb" /external:W1 /Gd /TP /errorReport:queue  -msse -msse2 -mcx16 "E:\\build-rpcs3-msvc\\CMakeFiles\\CMakeScratch\\TryCompile-szusic\\src.cxx"
        cl : command line  warning D9002: ignoring unknown option '-msse' [E:\\build-rpcs3-msvc\\CMakeFiles\\CMakeScratch\\TryCompile-szusic\\cmTC_a7f2c.vcxproj]
        cl : command line  warning D9002: ignoring unknown option '-msse2' [E:\\build-rpcs3-msvc\\CMakeFiles\\CMakeScratch\\TryCompile-szusic\\cmTC_a7f2c.vcxproj]
        cl : command line  warning D9002: ignoring unknown option '-mcx16' [E:\\build-rpcs3-msvc\\CMakeFiles\\CMakeScratch\\TryCompile-szusic\\cmTC_a7f2c.vcxproj]
          src.cxx
```
MSVC literally does not no these flags because. Only GCC and Clang support those.

Comment (Megamouse) [reply rpcs3/CMakeLists.txt:158]: I think you have to ifdef the check_cxx_compiler_flags to
`if(NOT WIN32 OR NOT MSVC)`
or something like that

Comment (digant73) [reply rpcs3/CMakeLists.txt:158]: but did you (oltolm) have the issue I had in llvm's `CMakeLists.txt` (the linking error on LLVMIntelJITEvents module)? Which CPU do you have?

File: CMakePresets.json (hunk 2 above)
Comment (digant73) [reply CMakePresets.json:51]: the `--config Debug`  and `--config Release` options seems to work from command line. I tested the standalone `cmake` tool (not the MSVC's cmake).
But how to set it from the IDE?

File: rpcs3/CMakeLists.txt (hunk 5 above)
Comment (oltolm) [reply rpcs3/CMakeLists.txt:158]: Yes I had the same issue, but it was fixed by this change
```
		if(LLVM_USE_INTEL_JITEVENTS)
			list (APPEND LLVM_ADDITIONAL_LIBS IntelJITEvents)
		endif()
		if(LLVM_USE_PERF)
			list (APPEND LLVM_ADDITIONAL_LIBS PerfJITEvents)
		endif()
```
`COMPILER_X86` is not defined for VC. You can check it yourself with a CMake debugger or just by adding this to `CMakeLists.txt`
```
message("COMPILER_X86: ${COMPILER_X86}")
```

Comment (digant73) [reply rpcs3/CMakeLists.txt:158]: > Yes I had the same issue, but it was fixed by this change
> 
> ```
> 		if(LLVM_USE_INTEL_JITEVENTS)
> 			list (APPEND LLVM_ADDITIONAL_LIBS IntelJITEvents)
> 		endif()
> 		if(LLVM_USE_PERF)
> 			list (APPEND LLVM_ADDITIONAL_LIBS PerfJITEvents)
> 		endif()
> ```
> 

yes I saw your other post. So with this solution we will link PerfJITEvents instead of IntelJITEvents. I don't know if that is ok or could be worse.

File: CMakePresets.json (hunk 2 above)
Comment (oltolm) [reply CMakePresets.json:51]: To set it from the IDE you need to define a build preset for each configuration. Something like this:
```
    "buildPresets": [
        {
            "name": "msvc-debug",
            "configurePreset": "msvc"
            "configuration": "Debug"
        },
        {
            "name": "msvc-release",
            "configurePreset": "msvc",
            "configuration": "Release"
        },
    ]
```

File: 3rdparty/llvm/CMakeLists.txt (hunk 1 above)
Comment (kd-11) [reply 3rdparty/llvm/CMakeLists.txt:21]: MSVC for arm isn't different than the one for X64 so it won't set COMPILER_ARM either.
The condition should just be `if(WIN32 AND (COMPILER_X86 OR MSVC))`. We don't support using msvc for arm64 windows building due to extensive problems with microsoft's compiler stack.

File: CMakePresets.json (hunk 2 above)
Comment (kd-11) [reply CMakePresets.json:51]: Verify that your sln and/or vcxproj file is not broken. All configured targets should appear in the dropdowns at the top for your CPU arch and build type.

Comment (oltolm) [reply CMakePresets.json:51]: He is not opening the "sln", he is opening the "rpcs3" directory.

**Comment by kd-11** [comment]
I've added myself as reviewer because I need to verify that windows-on-arm build doesn't break here. There is no CI for it and I have to spin up an expensive cloud instance to compile it, so I'll only do that once it is ready.

File: 3rdparty/llvm/CMakeLists.txt (hunk 1 above)
Comment (digant73) [reply 3rdparty/llvm/CMakeLists.txt:21]: yes, if COMPILER_X64 is not set on MSVC there are also few chances COMPILER_ARM will be set for that arch.

Comment (oltolm) [reply 3rdparty/llvm/CMakeLists.txt:21]: You should apply my changes in addition to that. If you look at the llvm cmake code you will see that `IntelJITEvents` is only built if `LLVM_USE_INTEL_JITEVENTS` is true and `PerfJITEvents` is only built if `LLVM_USE_PERF` is true.

File: CMakePresets.json (hunk 2 above)
Comment (digant73) [reply CMakePresets.json:51]: > To set it from the IDE you need to define a build preset for each configuration. Something like this:
> 
> ```
>     "buildPresets": [
>         {
>             "name": "msvc-debug",
>             "configurePreset": "msvc"
>             "configuration": "Debug"
>         },
>         {
>             "name": "msvc-release",
>             "configurePreset": "msvc",
>             "configuration": "Release"
>         },
>     ]
> ```

I saw that `buildPresets` were missing in the `CMakePresets.json`. I created instead two configPresets for `msvc-debug` and `msvc-release` but it didn't work (those have been added in the **Configuration** drop-down menu with `gcc` and `clang` and as I said both have been compiled in Debug mode for the reason you reported). I will add the `buildPresets`, they should not conflict with `gcc` and `clang` configs.

File: 3rdparty/llvm/CMakeLists.txt (hunk 1 above)
Comment (digant73) [reply 3rdparty/llvm/CMakeLists.txt:21]: > You should apply my changes in addition to that. If you look at the llvm cmake code you will see that `IntelJITEvents` is only built if `LLVM_USE_INTEL_JITEVENTS` is true and `PerfJITEvents` is only built if `LLVM_USE_PERF` is true.

Yes, I reported that in the other post in this thread. I already integrated your changes and kd-11 changes. Just making other checks and tests before I will push the changes

Comment (digant73) [reply 3rdparty/llvm/CMakeLists.txt]: fixed

File: CMakePresets.json (hunk 2 above)
Comment (digant73) [reply CMakePresets.json]: fixed

File: BUILDING.md (hunk 3 above)
Comment (digant73) [reply BUILDING.md]: fixed

File: rpcs3/CMakeLists.txt (hunk 5 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt]: fixed

File: rpcs3/CMakeLists.txt [hunk 6]
```diff
@@ -152,6 +152,11 @@ elseif(UNIX)
             COMMAND ${CMAKE_COMMAND} -E copy_directory
             ${CMAKE_SOURCE_DIR}/bin/git $<TARGET_FILE_DIR:rpcs3>/git)
 elseif(WIN32)
+    if(MSVC)
+        add_custom_command(TARGET rpcs3 POST_BUILD
+            COMMAND "${CMAKE_COMMAND}" -E copy "${PROJECT_BINARY_DIR}/3rdparty/OpenAL/openal-soft/$<CONFIG>/OpenAL32.dll" "$<TARGET_FILE_DIR:rpcs3>"
+            COMMAND "${CMAKE_COMMAND}" -E remove "${CMAKE_SOURCE_DIR}/bin/rpcs3.exe")
```
Comment (Megamouse) [inline rpcs3/CMakeLists.txt:158]: Why do you have to remove this?

File: rpcs3/CMakeLists.txt [hunk 7]
```diff
@@ -152,6 +152,11 @@ elseif(UNIX)
             COMMAND ${CMAKE_COMMAND} -E copy_directory
             ${CMAKE_SOURCE_DIR}/bin/git $<TARGET_FILE_DIR:rpcs3>/git)
 elseif(WIN32)
+    if(MSVC)
+        add_custom_command(TARGET rpcs3 POST_BUILD
+            COMMAND "${CMAKE_COMMAND}" -E copy "${PROJECT_BINARY_DIR}/3rdparty/OpenAL/openal-soft/$<CONFIG>/OpenAL32.dll" "$<TARGET_FILE_DIR:rpcs3>"
```
Comment (Megamouse) [inline rpcs3/CMakeLists.txt:157]: Why not do this in the OpenAL build scripts?

File: rpcs3/CMakeLists.txt (hunk 6 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt:158]: It is not needed to be remove it but only to exclude it in the copy made by the next command:
`COMMAND "${CMAKE_COMMAND}" -E copy_directory "${CMAKE_SOURCE_DIR}/bin" "$<TARGET_FILE_DIR:rpcs3>"`
But that command has no option to exclude files (in our case only `rpcs3.exe`) so decided to use a dirty solution removing it before the copy.
I could create an external cmake script to replace the `copy_directory` and make a copy excluding the rpcs3.exe file but is seems to me a complicated workaround. An alternative is also to rename the rpcs3.exe before the copy and then restore it after the copy. But even this solution seems to me horrible

File: rpcs3/CMakeLists.txt (hunk 7 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt:157]: I provided it here due to it is the place where the application build is finalized (all the resource files are copied here (even the external qt files)

File: rpcs3/CMakeLists.txt (hunk 6 above)
Comment (Megamouse) [reply rpcs3/CMakeLists.txt:158]: Why should it exclude the exe?

Comment (digant73) [reply rpcs3/CMakeLists.txt:158]: because `rpcs3.exe` is built under `rpcs3\build-msvc\bin` and the post build action, copying the `rpcs3\bin` in `rpcs3\build-msvc\bin`, will override the just compiled `rpcs3.exe` with the one in `rpcs3\bin`.
I'm looking for a solution more similar to the one used for UNIX where the override is avoided.

**Comment by digant73** [comment]
PR put on draft. Fixing some bugs on the post build action possibly due to the recent Qt update to 6.7.3

File: rpcs3/CMakeLists.txt (hunk 6 above)
Comment (digant73) [reply rpcs3/CMakeLists.txt]: Made a cleanup on post build actions for Windows on cmake solution. As for Unix and Apple, only needed resource files are copied from `rpcs3/bin` folder

File: 3rdparty/llvm/CMakeLists.txt (hunk 1 above)
Comment (digant73) [reply 3rdparty/llvm/CMakeLists.txt]: at the end the code needed for building on all supported OS and architecture was:

`if(WIN32 AND (COMPILER_X86 OR MSVC))`

and

```
if(LLVM_USE_INTEL_JITEVENTS OR (CMAKE_SYSTEM_NAME STREQUAL "Linux" AND COMPILER_X86))
	list (APPEND LLVM_ADDITIONAL_LIBS IntelJITEvents)
endif()
if(LLVM_USE_PERF OR CMAKE_SYSTEM_NAME STREQUAL "Linux")
	list (APPEND LLVM_ADDITIONAL_LIBS PerfJITEvents)
endif()

```

The check on `Linux` is needed for `gcc` and `clang`  due to the following options set in `CMakePresets.json`:

```
"BUILD_LLVM": "OFF",
"STATIC_LINK_LLVM": "ON"

```

With those options and without the check on `Linux`, libs `IntelJITEvents` (in particualr) and `PerfJITEvents` were not linked causing link errors such as the following:

```
ld.lld: error: undefined symbol: llvm::JITEventListener::createIntelJITEventListener()
>>> referenced by JITLLVM.cpp

```

**Comment by Megamouse** [comment]
why did you push spu code?

**Comment by digant73** [comment]
> why did you push spu code?

my mistake. I pushed some updates on this PR. thanks for the info. Reverting the change.
After that, this PR is ready for a review and merge. There is another incoming PR fixing and improving the SLN project. But it's better to provide it after the merge of this PR

File: CMakePresets.json [hunk 8]
```diff
@@ -7,12 +7,12 @@
       "binaryDir": "build-gcc",
       "cacheVariables": {
         "CMAKE_BUILD_TYPE": "Debug",
+        "USE_NATIVE_INSTRUCTIONS": "ON",
```
Comment (Megamouse) [inline CMakePresets.json:10]: Please stop changing unrelated things.
There's no reason to reorder these. All it does is add review overhead

Comment (digant73) [reply CMakePresets.json:10]: ok. I simply ordered all the 3 sections in the same way (they had almost all the same properties but in different order)
